pyomo/contrib/incidence_analysis/common/dulmage_mendelsohn.py

```diff
@@ -0,0 +1,101 @@
+#  ___________________________________________________________________________
+#
+#  Pyomo: Python Optimization Modeling Objects
+#  Copyright 2017 National Technology and Engineering Solutions of Sandia, LLC
+#  Under the terms of Contract DE-NA0003525 with National Technology and 
+#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain 
+#  rights in this software.
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
+
+from pyomo.common.dependencies import networkx_available
+
+if networkx_available:
+    from networkx.classes.digraph import DiGraph
+    from networkx.algorithms.traversal.breadth_first_search import bfs_edges
+    from networkx.algorithms.bipartite.basic import sets as bipartite_sets
+    from networkx.algorithms.bipartite.matching import maximum_matching
+    from networkx.algorithms.components import connected_components
+
+
+def _get_projected_digraph(bg, matching, top_nodes):
+    """
+    """
+    digraph = DiGraph()
+    digraph.add_nodes_from(top_nodes)
+    for n in top_nodes:
+        # Add in-edges
+        if n in matching:
+            for t in bg[matching[n]]:
+                if t != n:
+                    digraph.add_edge(t, n)
+        # Add out-edges
+        for b in bg[n]:
+            if b in matching and matching[b] != n:
+                digraph.add_edge(n, matching[b])
+    return digraph
+
+
+def _get_reachable_from(digraph, sources):
+    """
+    """
+    _filter = set()
+    reachable = []
+    for node in sources:
+        for i, j in bfs_edges(digraph, node):
```

Comment:
I don't think this buys us anything here, because `sources` only contains unmatched nodes, which are guaranteed to not be reachable, so they won't ever be encountered by a BFS. It could save work if we intend to use this function in other contexts, however.

Reply:
What _would_ buy us something is doing a single BFS with a set of nodes as the starting points. I did not see an easy way to do this with built-in NetworkX functionality short of adding a "common source" node pointing to all unmatched nodes and popping the node after the BFS, and did not think it would be worth the effort to implement this myself at this point. If we start seeing cases where the algorithm is slow because of this (i.e. many unmatched nodes, large over/underconstrained subsystems), we can explore a more intelligent BFS.